File: 744-network-delay-time/network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<pair<int,int>> adj[n+1];
        for(auto &it: times) {
            adj[it[0]].push_back({it[1], it[2]});
        }
        //adj list was done so create pq for iteration 
        vector<int> time(n+1, INT_MAX);
        time[k] = 0;
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        pq.push({0, k});

        while( !pq.empty() ) {
            int currTime = pq.top().first;
            int currNode = pq.top().second;
            pq.pop();

            if(currTime > time[currNode])
                continue;

            //iterate through the adjacent node using adj list
            for(auto it: adj[currNode]) {
                int adjNode = it.first;
                int extraTime = it.second;
                if(currTime + extraTime < time[adjNode]) {
                    pq.push({currTime + extraTime , adjNode});
                    time[adjNode] = currTime + extraTime;
                }
            }
        }

        int ans = INT_MIN; //cause we wanna find the max time it took in the shortest path among all ..
        for(int i = 1; i <= n; i++) {
            ans = max(ans, time[i]);
        }
        return (ans != INT_MAX) ? ans : -1;
    }
};

File: 744-network-delay-time/network-delay-time.java
class Solution {
    class Node {
        int node, time;
        public Node(int node, int time) {
            this.node = node;
            this.time = time;
        }
    }

    public int networkDelayTime(int[][] times, int n, int k) {
        int[] time = new int[n+1];
        for(int i = 1; i <= n; i++) {
            time[i] = Integer.MAX_VALUE;
        }

        List<Node>[] adj = new List[n+1];
        for(int i = 0; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }
        for(int[] it: times) {
            adj[it[0]].add(new Node(it[1], it[2]));
        }

        Queue<Node> queue = new PriorityQueue<>((n1, n2) -> n1.time - n2.time);
        queue.offer(new Node(k, 0));
        time[k] = 0;

        while( !queue.isEmpty() ) {
            Node it = queue.poll();
            int currNode = it.node;
            int currTime = it.time;

            for(int i = 0; i < adj[currNode].size(); i++) {
                Node nextPair = adj[currNode].get(i);//gets the ith node or pair
                int adjNode = nextPair.node;
                int extraTime = nextPair.time;
                if(currTime + extraTime < time[adjNode]) {
                    time[adjNode] = currTime + extraTime;
                    queue.add(new Node(adjNode, currTime + extraTime)); 
                }
            }
        }

        int ans = 0;
        for(int it: time) {
            ans = Math.max(ans, it);
        }
        return (ans != Integer.MAX_VALUE) ? ans : -1;
    }
}